src/signing/collector/signatures_collector.rs

```diff
@@ -0,0 +1,554 @@
+/// first; namely `LedgerFactorSource`, which user might also lack access to.
+pub struct SignaturesCollector {
+    /// Stateless immutable values used by the collector to gather signatures
+    /// from factor sources.
+    dependencies: SignaturesCollectorDependencies,
+
+    /// Mutable internal state of the collector which builds up the list
+    /// of signatures from each used factor source.
+    state: RefCell<SignaturesCollectorState>,
+}
+
+impl SignaturesCollector {
+    /// Used by our tests. But Sargon will typically wanna use `SignaturesCollector::new` and passing
+    /// it a
+    pub(crate) fn with(
+        all_factor_sources_in_profile: IndexSet<HDFactorSource>,
+        transactions: IndexSet<TXToSign>,
+        interactors: Arc<dyn SignatureCollectingInteractors>,
+    ) -> Self {
+        let preprocessor = SignaturesCollectorPreprocessor::new(transactions);
+        let (petitions, factors) = preprocessor.preprocess(all_factor_sources_in_profile);
+
+        let dependencies = SignaturesCollectorDependencies::new(interactors, factors);
+        let state = SignaturesCollectorState::new(petitions);
+
+        Self {
+            dependencies,
+            state: RefCell::new(state),
+        }
+    }
+
+    pub fn with_signers_extraction<F>(
+        all_factor_sources_in_profile: IndexSet<HDFactorSource>,
+        transactions: IndexSet<TransactionIntent>,
+        interactors: Arc<dyn SignatureCollectingInteractors>,
+        extract_signers: F,
+    ) -> Result<Self>
+    where
+        F: Fn(TransactionIntent) -> Result<TXToSign>,
+    {
+        let transactions = transactions
+            .into_iter()
+            .map(extract_signers)
+            .collect::<Result<IndexSet<TXToSign>>>()?;
+
+        let collector = Self::with(all_factor_sources_in_profile, transactions, interactors);
+
+        Ok(collector)
+    }
+
+    pub fn new(
+        transactions: IndexSet<TransactionIntent>,
+        interactors: Arc<dyn SignatureCollectingInteractors>,
+        profile: &Profile,
+    ) -> Result<Self> {
+        Self::with_signers_extraction(
+            profile.factor_sources.clone(),
+            transactions,
+            interactors,
+            |i| TXToSign::extracting_from_intent_and_profile(&i, profile),
+        )
+    }
+}
+
+// === PUBLIC ===
+impl SignaturesCollector {
+    pub async fn collect_signatures(self) -> SignaturesOutcome {
+        _ = self
+            .sign_with_factors() // in decreasing "friction order"
+            .await
+            .inspect_err(|e| eprintln!("Failed to use factor sources: {:#?}", e));
+
+        self.outcome()
+    }
+}
+
+// === PRIVATE ===
+impl SignaturesCollector {
+    /// If all transactions already would fail, or if all transactions already are done, then
+    /// no point in continuing.
+    ///
+    /// `Ok(true)` means "continue", `Ok(false)` means "stop, we are done". `Err(_)` means "stop, we have failed".
+    fn continue_if_necessary(&self) -> Result<bool> {
```

Comment:
TODO rename

Reply:
maybe something like `is_meaningful_to_continue` because `can` is too broad a term. We can always add signatures to a failed transaction, but it is not meaningful. 

Also, we can add signatures to an already valid transaction, but it is not meaningful.

Reply:
@GhenadieVP see: https://github.com/radixdlt/sargon-mfa/pull/3